### PR TITLE
Start leader based on Poh tick height.

### LINKED ISSUE #2570: the network stalls on leader failure
#### Problem
TVUs wait until they see all the data from the scheduled leader before starting their own, scheduled TPU.   
 There's no code to detect previous leader failure or timeout, no code to act or decide when to go ahead and transmit ticks for the previous leader's slot.

#### Proposed Solution
Construct a PohService at each vote point that includes a parameter that means "tick height when this poh should be a block (i.e. this node becomes leader).  Once this height is reached, this object can construct a TPU from the bank checkpoint the vote references.

### PR DESCRIPTION
**Pull request by aeyakovenko**
#### Problem

Cluster will stall if the previous slot fails.  With this change leader start is triggered by poh.

#### Summary of Changes

* look at current Poh tick height
* if this node is the current leader based on tick height, start the bank
* test that cluster runs if leader is killed

depends on #3083 #3082 
Fixes #2570

Not quite there yet :)

failures:

---- test_leader_failure_3 stdout ----
thread 'test_leader_failure_3' panicked at 'called `Result::unwrap()` on an `Err` value: Custom { kind: Other, error: StringError("retry_transfer failed") }', src/libcore/result.rs:1009:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.

#### Debugging Notes
to run:
```
RUST_LOG=solana::fullnode=debug,solana::poh_recorder=info,solana::banking_stage=debug,solana::replay_stage=debug,solana::cluster_tests=trace
cargo watch -x 'test --all leader_failure_2 -- --nocapture'
```

leader never changes for any slot
```
[2019-03-04T19:22:38.509781355Z DEBUG solana::replay_stage] me: 8HXZRTxh3b2KsxsrieimbUCkCeBF41kJ61P2uX9oSUGa leader BsEGNV5Uhu2u8Gn6vkjyFmkdtJwPVSPY1Fwbgmoy738b at poh slot 32
[2019-03-04T19:22:38.612110450Z DEBUG solana::replay_stage] me: 8HXZRTxh3b2KsxsrieimbUCkCeBF41kJ61P2uX9oSUGa leader BsEGNV5Uhu2u8Gn6vkjyFmkdtJwPVSPY1Fwbgmoy738b at poh slot 33
```

### CODE REVIEW

File: core/src/poh_recorder.rs [hunk 1]
```diff
@@ -425,4 +433,60 @@ mod tests {
...
+        poh_recorder.reset(
+            poh_recorder.tick_cache[0].1,
+            poh_recorder.tick_cache[0].0.hash,
+        );
+        assert_eq!(poh_recorder.tick_cache.len(), 2);
+        poh_recorder.reset(
+            poh_recorder.tick_cache[1].1,
+            poh_recorder.tick_cache[1].0.hash,
+        );
+        assert_eq!(poh_recorder.tick_cache.len(), 2);
```
Comment (rob-solana) [inline core/src/poh_recorder.rs]: how's this still 2?  I guess I don't understand what reset() does.

File: core/src/replay_stage.rs [hunk 2]
```diff
@@ -201,7 +168,46 @@ impl ReplayStage {
...
+            let poh_tick_height = poh_recorder.lock().unwrap().tick_height();
+            let poh_slot = leader_schedule_utils::tick_height_to_slot(parent, poh_tick_height + 1);
+            assert!(frozen.get(&poh_slot).is_none());
+            if bank_forks.read().unwrap().get(poh_slot).is_none() {
+                let next_leader = leader_schedule_utils::slot_leader_at(poh_slot, parent);
+                cluster_info.write().unwrap().set_leader(next_leader);
+                if next_leader == my_id {
+                    let tpu_bank = Bank::new_from_parent(parent, my_id, poh_slot);
+                    bank_forks.write().unwrap().insert(poh_slot, tpu_bank);
+                    if let Some(tpu_bank) = bank_forks.read().unwrap().get(poh_slot).cloned() {
```
Comment (rob-solana) [inline core/src/replay_stage.rs]: why might this fail?

File: core/src/poh_recorder.rs (hunk 1 above)
Comment (aeyakovenko) [reply core/src/poh_recorder.rs]: That value exists in the cache.  So there is nothing to reset.

File: core/src/rpc.rs [hunk 3]
```diff
@@ -87,6 +108,20 @@ impl JsonRpcRequestProcessor {
...
+
+    pub fn fullnode_exit(&self) -> Result<bool> {
+        match self.config {
+            JsonRpcConfig::Safe => {
+                debug!("safe config, fullnode_exit ignored");
+                Ok(false)
+            }
+            JsonRpcConfig::Unsafe => {
+                warn!("JsonRPC fullnode_exit request...");
+                self.fullnode_exit.store(true, Ordering::Relaxed);
```
Comment (rob-solana) [inline core/src/rpc.rs]: want to guard this with a command line option?

File: core/src/cluster_tests.rs [hunk 4]
```diff
@@ -39,3 +41,56 @@ pub fn spend_and_verify_all_nodes(
...
+    assert!(cluster_nodes.len() >= nodes);
+    let mut client = mk_client(&entry_point_info);
+    assert!(client.fullnode_exit().unwrap());
+    for ingress_node in &cluster_nodes {
+        if ingress_node.id == entry_point_info.id {
+            continue;
+        }
+        let random_keypair = Keypair::new();
+        let mut client = mk_client(&ingress_node);
+        let bal = client
```
Comment (pgarg66) [inline core/src/cluster_tests.rs]: Maybe worth refactoring this function and ```spend_and_verify_all_nodes()```. There's a lot of common code.

File: core/src/replay_stage.rs [hunk 5]
```diff
@@ -158,28 +144,9 @@ impl ReplayStage {
...
-                                    my_id,
-                                    tpu_bank.slot(),
-                                    next_leader
-                                );
-                				poh_recorder.lock().unwrap().set_bank(&tpu_bank);
-                            }
-                        }
                     }
+
+                    Self::start_leader(my_id, &bank_forks, &poh_recorder, &cluster_info);
```
Comment (carllin) [inline core/src/replay_stage.rs]: if the replay_blocktree_into_bank()  processing above takes a long time, can the poh_recorder potentially tick past the slot for which this node is supposed to be the leader?

Comment (aeyakovenko) [reply core/src/replay_stage.rs]: Yea. Unless that block completes and the poh is reset

**Pull request merged**